inspirehep/modules/workflows/tasks/matching.py

```diff
@@ -446,14 +446,50 @@ def _get_wfs_same_source(obj, eng):
 
         for wf_id in workflows:
             wf = workflow_object_class.get(wf_id)
-            wf_source = get_value(wf.data, 'acquisition_source.source').lower()
-            if wf_source == current_source:
+            if _has_same_source(current_source, wf):
                 return True
         return False
 
     return _get_wfs_same_source
 
 
+def _has_same_source(current_wf_source, holdingpen_match_wf):
+    wf_source = get_value(holdingpen_match_wf.data, 'acquisition_source.source').lower()
+    if wf_source == current_wf_source:
+        return True
+
+
+def _replace_matched_wf_steps_with_stop_steps(stopping_steps, matched_wf_eng, matched_wf):
+    matched_wf_eng.callbacks.replace(stopping_steps)
+    matched_wf_eng.process([matched_wf])
+
+
+def _halt_record_if_match_from_different_source(holdingpen_wf_eng, holdingpen_wf):
+    try:
+        holdingpen_wf_eng.halt(msg="Waitning for matched worfklow to finish")
+    except HaltProcessing:
+        holdingpen_wf.status = ObjectStatus.HALTED
+        holdingpen_wf.extra_data['halted-by-match-with-different-source'] = True
+        holdingpen_wf_eng.save()
+        holdingpen_wf.save()
+
+
+def _handle_matched_holdingpen_wfs(obj, same_source_only=False):
+    current_wf_source = get_value(obj.data, 'acquisition_source.source').lower()
+    stopping_steps = [mark('stopped-by-wf', int(obj.id)), stop_processing]
+    for holdingpen_wf_id in obj.extra_data['holdingpen_matches']:
+        holdingpen_wf = workflow_object_class.get(holdingpen_wf_id)
+        holdingpen_wf_eng = WorkflowEngine.from_uuid(holdingpen_wf.id_workflow)
+
+        # halt workflow
+        if same_source_only and not _has_same_source(current_wf_source, holdingpen_wf):
+            _halt_record_if_match_from_different_source(holdingpen_wf_eng, holdingpen_wf)
+            continue
+
+        # stop this holdingpen workflow by replacing its steps with a stop step
+        _replace_matched_wf_steps_with_stop_steps(stopping_steps, holdingpen_wf_eng, holdingpen_wf)
```

Comment:
is this just used here? Shall we just add the two lines here instead of a separate function?